libass/ass_cache.c

```diff
@@ -375,87 +398,131 @@ void *ass_cache_get(Cache *cache, void *key, void *priv)
 {
     const CacheDesc *desc = cache->desc;
     size_t key_offs = CACHE_ITEM_SIZE + align_cache(desc->value_size);
-    unsigned bucket = desc->hash_func(key, ASS_HASH_INIT) % cache->buckets;
+    ass_hashcode hash = desc->hash_func(key, ASS_HASH_INIT);
+    unsigned bucket = hash % cache->buckets;
+
+    CacheItem * _Atomic *bucketptr = &cache->map[bucket];
+    CacheItem *stop_at = NULL;
+    CacheItem *item, *new_item = NULL;
+    void *new_key = NULL;
+    CacheItem *start = *bucketptr;
+
+retry:
+    for (item = start; item && item != stop_at; item = item->next) {
+        if (item->hash == hash && desc->compare_func(key, (char *)item + key_offs))
+            break;
+    }
 
-    pthread_mutex_lock(&cache->mutex);
+    if (item != NULL && item != stop_at) {
+        atomic_store_explicit(&item->last_used_frame, cache->cur_frame, memory_order_relaxed);
 
-    CacheItem *item = cache->map[bucket];
-    while (item) {
-        if (desc->compare_func(key, (char *) item + key_offs)) {
-            if (!item->queue_prev || item->queue_next) {
-                if (item->queue_prev) {
-                    item->queue_next->queue_prev = item->queue_prev;
-                    *item->queue_prev = item->queue_next;
-                } else
-                    inc_ref(&item->ref_count);
-                *cache->queue_last = item;
-                item->queue_prev = cache->queue_last;
-                cache->queue_last = &item->queue_next;
-                item->queue_next = NULL;
-            }
-            cache->hits++;
+        atomic_fetch_add_explicit(&cache->hits, 1, memory_order_relaxed);
+
+        if (new_item && desc->key_destruct_func)
```

Comment:
`key_move_func` potentially being called twice is deliberate (once from `item` to `new_item` and once from `new_item` to `NULL`).

Reply:
I feel that it's logically incorrect to use `key_move_func()` in place of `key_destruct_func()`. There're two forms of keys with different ownership properties: request key (`ass_cache_get()` argument) and item key. So, `key_move_func()` is about destroying request key and using it for item key looks like a hack.

Maybe it's better to make `key_destruct_func()` mandatory and `destruct_func()` do value-only destruction?

Reply:
`key_move_func` is called when we're releasing a request key, while `key_destruct_func` is called when we're destroying an item key on an unconstructed item. Cases where `key_destruct_func` isn't set mean that `key_move_func(NULL, key)` will do the same thing it would've done. This is all kind of awkward since ownership and lifetimes on key contents depends on the item type.